addons/project/models/project.py

```diff
@@ -732,6 +737,21 @@ def _read_group_stage_ids(self, stages, domain, order):
     repeat_show_week = fields.Boolean(compute='_compute_repeat_visibility')
     repeat_show_month = fields.Boolean(compute='_compute_repeat_visibility')
 
+    def _is_cyclic_dependency(self):
+        self.ensure_one()
+        dependent_tasks = self.depend_on_ids
+        while dependent_tasks:
+            if self.id in dependent_tasks.ids:
+                return True
+            dependent_tasks = dependent_tasks.mapped('depend_on_ids')
+        return False
+
+    @api.constrains('depend_on_ids')
+    def _check_no_cyclic_dependencies(self):
+        tasks_with_dependency = self.filtered(lambda task: task.depend_on_ids)
+        if any(task._is_cyclic_dependency() for task in tasks_with_dependency):
+            raise ValidationError("You cannot create cyclic dependency.")
```

Comment:
You should use the builtin helper `_check_m2m_recursion` (defined in `models.py`)

Reply:
Cool, I didn't know that, thanks !